openmls/src/treesync/node/codec.rs

```diff
@@ -112,7 +116,22 @@ impl tls_codec::Deserialize for ParentNode {
     {
         let public_key = HpkePublicKey::tls_deserialize(bytes)?;
         let parent_hash = VLBytes::tls_deserialize(bytes)?;
-        let unmerged_leaves = Vec::tls_deserialize(bytes)?;
+        let unmerged_leaves = UnmergedLeaves::tls_deserialize(bytes)?;
+
         Ok(Self::new(public_key, parent_hash, unmerged_leaves))
     }
 }
+
+impl tls_codec::Deserialize for UnmergedLeaves {
+    fn tls_deserialize<R: std::io::Read>(bytes: &mut R) -> Result<Self, tls_codec::Error>
+    where
+        Self: Sized,
+    {
+        let list = Vec::tls_deserialize(bytes)?;
+        Self::try_from(list).map_err(|e| match e {
```

Comment:
I'd generally like to get rid and void any additional deserialization failures that are not due to the serialization.
It's of course some tradeoff because it requires an additional call on the outside and some wrapper struct(s).
I'd be fine with this if you don't want to change it. But we should probably do some nice helpers for something like this (e.g. verify functions that can be called during deserialization) and make use of them here.

Reply:
Right, we have this in a few places now. Happy to discuss this if you think combining this is problematic. Personally, I'm not too worried about this, as long as the extra validation is happening separately and not in the codec part of the code. 
I created #1132 to that effect.